File: js/test/literal_text.test.js
import { test } from 'node:test';
import assert from 'node:assert/strict';
import { readFileSync } from 'node:fs';
import { validateArguments } from '../src/contracts.js';
import { validateArg, validateArgWithCustomTypes } from '../src/validator.js';
const cases = JSON.parse(readFileSync(new URL('../../tests/literal_text_vectors.json', import.meta.url))).cases;
for (const c of cases) for (const custom of [false, true]) test(`literal_text ${c.name} custom=${custom}`, () => {
  const value = c.value.repeat(c.repeat);
  const def = {type: 'literal_text', required: true, ...(c.pattern !== null ? {pattern: c.pattern} : {})};
  const check = () => custom ? validateArgWithCustomTypes({...def, type:'source_text'}, value,
    {source_text: {base:'literal_text', ...(c.pattern !== null ? {pattern:c.pattern} : {})}}) : validateArg(def, value);
  if (c.error) assert.throws(check); else assert.equal(check(), value);
});
for (const value of ['\ud800', '\udfff', 1, false, null, {}, []]) test(`literal_text refuses ${JSON.stringify(value)}`, () => {
  assert.throws(() => validateArg({type:'literal_text'}, value));
  assert.throws(() => validateArgWithCustomTypes({type:'source_text'}, value, {source_text:{base:'literal_text'}}));
});

for (const value of [null, undefined, 1, false, {}, []]) test(`literal_text contract refuses ${String(value)}`, () => {
  const manifest = {args:{content:{type:'literal_text',default:'fallback'}}};
  assert.throws(() => validateArguments(manifest, {content:value}));
});

File: go/pkg/validator/literal_text_test.go
package validator

import (
	"encoding/json"
	"github.com/thirdkeyai/toolclad/pkg/manifest"
	"os"
	"strings"
	"testing"
)

func TestLiteralTextVectors(t *testing.T) {
	data, err := os.ReadFile("../../../tests/literal_text_vectors.json")
	if err != nil {
		t.Fatal(err)
	}
	var vectors struct {
		Cases []struct {
			Name, Value, Pattern string
			Repeat               int
			Error                bool
		}
	}
	if err := json.Unmarshal(data, &vectors); err != nil {
		t.Fatal(err)
	}
	if len(vectors.Cases) != 12 {
		t.Fatal("incomplete vectors")
	}
	for _, c := range vectors.Cases {
		t.Run(c.Name, func(t *testing.T) {
			value := strings.Repeat(c.Value, c.Repeat)
			def := &manifest.ArgDef{Name: "value", Type: "literal_text", Required: true, Pattern: c.Pattern}
			alias := &manifest.ArgDef{Name: "value", Type: "source_text", Required: true}
			types := map[string]*manifest.CustomTypeDef{"source_text": {Base: "literal_text", Pattern: c.Pattern}}
			for _, custom := range []bool{false, true} {
				var got string
				var err error
				if custom {
					got, err = ValidateArgWithCustomTypes(alias, value, types)
				} else {
					got, err = ValidateArg(def, value)
				}
				if c.Error {
					if err == nil {
						t.Fatal("expected refusal")
					}
				} else if err != nil || got != value {
					t.Fatalf("text changed or refused: %v", err)
				}
			}
		})
	}
}
func TestLiteralTextInvalidUTF8(t *testing.T) {
	for _, value := range []string{string([]byte{0xff}), string([]byte{0xed, 0xa0, 0x80})} {
		if _, err := ValidateArg(&manifest.ArgDef{Type: "literal_text"}, value); err == nil {
			t.Fatal("accepted invalid UTF-8")
		}
	}
}
